Support markdown when using @Description Annotation

cnabro opened the pull request:
Motivation:
- Support markdown when using @Description Annotation
- Armeria DocService can be extended and used for various purposes

Modifications:
- https://github.com/line/armeria/issues/4299#issue-1269566872
- [x] Support varius markup
  - markdown 
    - react-markdown
    - react-syntax-highlighter
    - remark-gfm
    - github-markdown-css
  - mermaid.js
- [x] Support enums documentation in annotated service.
- [ ] ~~Support return types documentation in annotated service using annotation.~~
  - ~~Add `@ResponseObject` to specify which element should be converted to structs in `DocService` by `AnnotatedDocServicePlugin`~~ 
  -  ~~Need to be exclude default class like `HttpResponse` ..~~
  - 👉 Rollbacked by https://github.com/line/armeria/issues/4309, it will be fixed using `StructInfoProvider`

- [x] Show path mappings in `Services`
- [x] Update Armeria Reference doc : https://armeria.dev/docs/server-docservice/

<img width=1681" alt="image" src="https://user-images.githubusercontent.com/6916639/174469675-daa21c8a-6fd3-40c0-bd78-0c538bdfcf0c.png">
<img width="1681" alt="image" src="https://user-images.githubusercontent.com/6916639/174469687-10c0c992-7c0f-4dde-81fb-63884bd6bc2a.png">
<img width="1671" alt="image" src="https://user-images.githubusercontent.com/6916639/174469701-ebaf22fb-fe15-48d1-a878-29a0b1b68c0c.png">


Result:
- Closes [#4299, #2200]
- Support Markdown Description 
- Support Enums&Return Types documentation in annotated service

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/line/armeria?pullRequest=4304) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/line/armeria?pullRequest=4304) before we can accept your contribution.<br/><hr/>**gyuil-han** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/line/armeria?pullRequest=4304) it.</sub>

minwoox commented on docs-client/src/lib/specification.tsx:
```diff
@@ -242,23 +235,8 @@ export class Specification {
     }
 
     // TODO(trustin): Handle the docstrings of enum values.
```
Let's also remove this comment.

minwoox commented on site/src/pages/docs/server-docservice.mdx:
```diff
@@ -211,18 +211,44 @@ files into the specific location in the class path.
+  import com.linecorp.armeria.server.annotation.ResponseObject;
 
-  @Description("A service that provides user information.")
+  @Description("""
```
This is Java so we cannot use the String template. 😭

cnabro replied:
Ah, I forgot this specification not merged yet 👀 (https://openjdk.org/jeps/8273943).
I'll fix it 😅

minwoox reviewed (commented):
This looks really amazing. Thanks a lot, @cnabro! 😄 
Do you mind adding test cases of `ResponseObject` and `Description` to Java test class especially to `AnnotatedDocServiceTest`?
We use the class as a playground of DocService using the env value (`DOC_SERVICE_DEMO`) so it would be very helpful if we can use it there. 😄

cnabro commented:
@minwoox ok, i'll try it. Thanks for the review ! 👍

ikhoon commented on core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedDocServicePlugin.java:
```diff
@@ -371,6 +374,15 @@ static TypeSignature toTypeSignature(Type type) {
+            return TypeSignature.ofNamed(clazz);
+        }
+
+        final ResponseObject responseObject = AnnotationUtil.findFirst(clazz, ResponseObject.class);
```
What do you think of using `StructInfoProvider` to determine whether to convert a return type to a named `TypeSignature`? #4309

cnabro replied:
Did you mean that add entire return type classes to named `TypeSignature` and then filter using `StructInfoProvider`?
e.g. )
```java
    private static NamedTypeInfo newNamedTypeInfo(TypeSignature typeSignature, @Nullable StructInfoProvider structInfoProvider) {
        final Class<?> type = (Class<?>) typeSignature.namedTypeDescriptor();
        if (type == null) {
            throw new IllegalArgumentException("cannot create a named type from: " + typeSignature);
        }

        if (type.isEnum()) {
            return newEnumInfo(type);
        }

        if (structInfoProvider != null) {
            const StructInfo structInfo = structInfoProvider.newStructInfo(type);
            if (structInfo != null) {
                return structInfo;
            }
        }

        return newStructInfo(type);
    }

```

I had similar concerns too about return type to a named `TypeSignature`.

first, I added entire return type classes using `TypeSignature.ofNamed(clazz)`.
But It had simillar problem like #4309. (need to be exclude default class like `HttpResponse`, `Date`, `HttpResult`.)
So I added annotation like `@ResponseObject` and filter it. 😅 

I think It's a good way to solve.

cnabro replied:
OK. I'll fix it contain https://github.com/line/armeria/issues/4309. 😄

minwoox replied:
Unless you want to remove `ResponseObject` and only use the `StructInfoProvider`, I think we can implement it separately in another PR.

cnabro commented:
As mentioned in Slack, 
Rollback `@ResponseObject` for #4309. And I added some test-case for AnnotatedDocServiceTest.

@minwoox @ikhoon

minwoox commented:
<img width="591" alt="image" src="https://user-images.githubusercontent.com/25103250/175313035-48db623d-84e9-4b7c-9e95-ee0ad38097c6.png">
I found out that the markdown String is not properly rendered when the page has a scroll.
We need to fix that or we can probably just disable markdown for the parameter and struct description.

minwoox commented:
> And I found the same issue on previous version (on disabled markdown). I think it's not a markdown issue

Oops, I didn't know that we had such a bug. 😓 

> I think the workaround is not use table

Then which one do you want to use instead? Could you tell me more about this, please? 😄

cnabro commented:
@minwoox 
After write comments, I found out some solution. 
Wrapping `<Table>` with `<TableContainer>` will be adjust `overflowX`

https://github.com/mui/material-ui/blob/v4.9.0/packages/material-ui/src/TableContainer/TableContainer.js#L10

How about fixing the issue like this screenshot using scroll?
![image](https://user-images.githubusercontent.com/6916639/175445816-a51074b5-4893-4a79-9521-d76a243c017b.png)

minwoox commented:
> How about fixing the issue like this screenshot using scroll?

That looks awesome. 😆

minwoox commented:
Had a chat with @trustin and how about adding a metadata to `@Description` so that we can support various markup such as [Mermaid.js](https://mermaid-js.github.io/mermaid/#/)?
```java
@Description(value= "...", markup = Markup.NONE)
@Description(value= "...", markup = Markup.MARKDOWN)
@Description(value= "...", markup = Markup.MERMAID) // we can add this later.
```

cnabro commented:
@minwoox Good idea. It will be used for various purposes. 👍

cnabro commented:
I modified `@Description` and add markup metadata to all types.
Now possible enable/disable markup using markup flags.

And I also added `mermaid.js` 😎
It will be rendered like this when using `Markup.MERMAID`
![image](https://user-images.githubusercontent.com/6916639/175657582-557f82ef-16ea-4cb5-a366-56c325051fb4.png)

minwoox reviewed (commented):
I didn't expect that you also implement `Mermaid.js`. 👍 👍 👍

cnabro commented on site/src/pages/docs/server-docservice.mdx:
````diff
@@ -211,18 +211,40 @@ files into the specific location in the class path.
 
   ```java
   import com.linecorp.armeria.server.annotation.Description;
+  import com.linecorp.armeria.server.annotation.ResponseObject;
````
I think's it's already removed 👀

minwoox replied:
Oops, sorry. I think it's a bug in my browser. 😅 
I don't know why but it sometimes caches the previous comment and adds it again.

cnabro commented on core/src/main/java/com/linecorp/armeria/server/docs/DocService.java:
```diff
@@ -122,24 +124,24 @@ public static DocServiceBuilder builder() {
         this(/* exampleHeaders */ ImmutableMap.of(), /* exampleRequests */ ImmutableMap.of(),
-             /* examplePaths */ ImmutableMap.of(), /* exampleQueries */ ImmutableMap.of(),
-             /* injectedScriptSuppliers */ ImmutableList.of(), DocServiceBuilder.ALL_SERVICES);
+                /* examplePaths */ ImmutableMap.of(), /* exampleQueries */ ImmutableMap.of(),
```
Ok. I just import `LINE OSS` using Eclipse XML profile. It is different from origin source.
So I fixed lints manually 🥲 ; Thank you for guiding this ! 
![image](https://user-images.githubusercontent.com/6916639/175873733-7e3e5deb-38fc-4f46-9678-39d4f5048b46.png)

cnabro commented on core/src/main/java/com/linecorp/armeria/server/annotation/DescriptionInfo.java:
```diff
@@ -0,0 +1,53 @@
+    /**
+     * Returns the documentation string.
+     */
+    public String getDocString() {
```
Ok. Thank you for guiding.

cnabro commented on core/src/main/java/com/linecorp/armeria/server/annotation/DescriptionInfo.java:
```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright 2018 LINE Corporation
```
I fixed it to `2022`

minwoox commented on core/src/main/java/com/linecorp/armeria/server/docs/DocService.java:
```diff
@@ -254,65 +265,97 @@ private static MethodInfo addMethodDocStrings(ServiceInfo service, MethodInfo me
+                             descriptionInfo.markup());
     }
 
     @Nullable
```
This doesn't return null value. 😉

minwoox reviewed (commented):
Thanks! Left a few minor comments. 😉 
It really looks good. 😄

cnabro commented on core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedDocServicePlugin.java:
```diff
@@ -162,10 +166,14 @@ private static void addMethodInfo(Map<Class<?>, Set<MethodInfo>> methodInfos,
                             name, returnTypeSignature, fieldInfos, ImmutableList.of(), // Ignore exceptions.
-                            ImmutableList.of(endpoint), httpMethod, AnnotatedServiceFactory
-                                    .findDescription(method));
+                            ImmutableList.of(endpoint), httpMethod, docString, supportedMarkup);
```
This is the part I was also thought about as well.
I thought the changes were smaller than using just one `DescriptionIfno` with two parameters.

But I can change it. 😅

cnabro commented on core/src/main/java/com/linecorp/armeria/server/docs/EnumValueInfo.java:
```diff
@@ -100,6 +104,15 @@ public String docString() {
+     */
+    @JsonProperty
+    @JsonInclude(Include.NON_NULL)
+    public Markup supportedMarkup() {
```
It will be changed to `DescriptionInfo`

cnabro commented on docs-client/src/components/Description/index.tsx:
```diff
@@ -0,0 +1,92 @@
+  if (!docString) {
+    return <div />;
+  }
+
```
I missed it. Thank you.

cnabro commented on docs-client/src/lib/specification.tsx:
```diff
@@ -240,25 +240,8 @@ export class Specification {
-    this.data.structs.forEach(this.renderDocString);
   }
 
   private updateStructDocStrings(structs: Struct[]) {
```
@minwoox I think `parseParamDocStrings`,`updateStructDocStrings` and `updateDocStrings` also can be removed (Because of the `Description` component). What do you think of that?

minwoox replied:
I think we need that. 😅 
@jrhee17 will leave a comment about this, so please hold a minute. 😉

jrhee17 replied:
Thanks @minwoox  😅  Sorry about the late comment.

I think the previous behavior was as follows:
> If a javadoc exists, then extract the `@param` values and inject these values as the parameters.

I think this feature is useful, especially for use-cases where teams want to document both code-level and `DocService` based on javadocs only. Also, I believe `@Description` is more explicit, and should have a higher priority than the `javadoc @param` based method.

What do you think of modifying the priority such that..
1. If a `@Description.docString` for a parameter/field doesn't exist, check the method/struct for `javadoc @param` and fill it if possible.
2. Otherwise, just use the original `docString`.

Additional notes:
@minwoox also suggested this logic can be moved to the java side in the long term (and I agree with this idea).
I think a separate PR can be uploaded as a follow-up, since I think the above logic change isn't too complex.

cnabro replied:
Thank you for a detailed explanation.
I didn't expected javaDoc specification.
I also agreed `this logic can be moved to the java side in the long term`.

First, I will rollback removed code `parseParamDocStrings`,`updateStructDocStrings` and `updateDocStrings`.
And then, I will send separate PR 👍

jrhee17 reviewed (commented):
Overall looks really nice 👍  ! Mostly nits for me 🙇 🙇 🙇

cnabro commented on core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedDocServicePlugin.java:
```diff
@@ -162,10 +166,14 @@ private static void addMethodInfo(Map<Class<?>, Set<MethodInfo>> methodInfos,
         final Class<?> clazz = service.object().getClass();
         route.methods().forEach(
                 httpMethod -> {
+                    final DescriptionInfo docStringInfo = AnnotatedServiceFactory.findDescription(method);
```
I think it's already fixed 😅
```java
route.methods().forEach(
        httpMethod -> {
            final MethodInfo methodInfo = new MethodInfo(
                    name, returnTypeSignature, fieldInfos, ImmutableList.of(), // Ignore exceptions.
                    ImmutableList.of(endpoint), httpMethod,
                    AnnotatedServiceFactory.findDescription(method));

            methodInfos.computeIfAbsent(clazz, unused -> new HashSet<>()).add(methodInfo);
        });
```

jrhee17 replied:
A lot of these comments were from yesterday before the PR update 😅   Feel free to resolve if already addressed 👍

cnabro commented on core/src/main/java/com/linecorp/armeria/server/docs/ServiceInfo.java:
```diff
@@ -77,12 +89,14 @@ public ServiceInfo(String name,
         this.methods = mergeEndpoints(requireNonNull(methods));
         this.exampleHeaders = ImmutableList.copyOf(requireNonNull(exampleHeaders, "exampleHeaders"));
         this.docString = Strings.emptyToNull(docString);
+        this.supportedMarkup = markup;
```
It's also fixed in https://github.com/line/armeria/pull/4304/commits/24addb600be64de99d1e8a113e92f593dd58475a

cnabro commented on core/src/main/java/com/linecorp/armeria/server/docs/FieldInfo.java:
```diff
@@ -77,18 +89,21 @@ public static FieldInfoBuilder builder(String name, TypeSignature typeSignature,
         this.typeSignature = typeSignature;
         this.childFieldInfos = childFieldInfos;
         this.docString = docString;
+        this.supportedMarkup = supportedMarkup;
```
It's also fixed in https://github.com/line/armeria/pull/4304/commits/24addb600be64de99d1e8a113e92f593dd58475a

cnabro commented on core/src/main/java/com/linecorp/armeria/server/docs/DocServicePlugin.java:
```diff
@@ -62,7 +63,8 @@ public interface DocServicePlugin {
 
     /**
-     * Loads the documentation strings that describes services and their methods, enums and their values and
+     * Loads the documentation description infos include strings and markup types
```
Thank you for fixing it 😎

cnabro commented on core/src/test/java/com/linecorp/armeria/internal/server/annotation/AnnotatedDocServiceTest.java:
```diff
@@ -157,170 +162,206 @@ void jsonSpecification() throws InterruptedException {
         final EndpointInfo endpoint = EndpointInfo.builder("*", "exact:/service/allMethods")
-                                                  .availableMimeTypes(MediaType.JSON_UTF_8)
-                                                  .build();
+                .availableMimeTypes(MediaType.JSON_UTF_8)
```
It's already fixed 👍

cnabro commented on docs-client/src/components/Description/Mermaid/index.tsx:
```diff
@@ -0,0 +1,57 @@
+ */
+
+import React, { useEffect } from 'react';
+// @ts-ignore
```
Ok. I'll try 👍

cnabro replied:
Good!. `@types/mermaid` is better then `mermaid`.
It will be changed like this
```
const Mermaid: React.FunctionComponent<MermaidProps> = ({ chart }) => {
  useEffect(() => {
    mermaid.initialize({
      theme: Theme.Base,
      logLevel: LogLevel.Fatal,
      securityLevel: SecurityLevel.Strict,
      startOnLoad: true,
      arrowMarkerAbsolute: false,
    });

    mermaid.contentLoaded();
  });

  if (!chart) {
    return <div />;
  }

  return <div className="mermaid">{chart}</div>;
};

cnabro commented on site/src/pages/docs/server-docservice.mdx:
````diff
@@ -212,17 +212,42 @@ files into the specific location in the class path.
+  }
   ```
 
+  - Support enum type descriptions in annotated service.
````
I removed this line. Thank you 😎

minwoox commented on docs-client/src/components/Description/Mermaid/index.tsx:
```diff
@@ -0,0 +1,57 @@
+    arrowMarkerAbsolute: false,
+  };
+
+  mermaid.initialize({ ...configs });
```
This is the related one: https://github.com/mermaid-js/mermaid/issues/311

minwoox commented on site/src/pages/docs/server-docservice.mdx:
````diff
@@ -212,17 +212,41 @@ files into the specific location in the class path.
+  }
   ```
 
+  - Support markdown ([GitHub Flavored Markdown Spec](https://github.github.com/gfm/))
+  - Support mermaid.js ([mermaid.js](https://mermaid-js.github.io/mermaid/#/))
````
How about just:
```suggestion
  - As you noticed, you can render [Markdown](https://github.github.com/gfm/) or [Mermaid](https://mermaid-js.github.io/mermaid/#/) using <type://Markup>.
```

minwoox reviewed (approved):
You've beautifully made it! Thank you! 😄

cnabro commented:
@minwoox @jrhee17 @ikhoon
Thanks for the review 👍

jrhee17 reviewed (approved):
Only left nits! Looks awesome 🙇 Thanks @cnabro 👍

cnabro commented on core/src/main/java/com/linecorp/armeria/server/docs/StructInfo.java:
```diff
@@ -72,9 +74,14 @@ public List<FieldInfo> fields() {
+    /**
+     * Returns the description information of the struct.
+     */
+    @JsonProperty
```
👍

cnabro commented on core/src/main/java/com/linecorp/armeria/server/annotation/DescriptionInfo.java:
```diff
@@ -0,0 +1,84 @@
+ * under the License.
+ */
+
+package com.linecorp.armeria.server.annotation;
```
OK. I'll fix it 👍

ikhoon reviewed (approved):
Awesome, @cnabro! 🚀💯
As you know, documentation service is the most beloved feature.
Many Armeria users will be happy with this.

cnabro commented on thrift0.13/src/main/java/com/linecorp/armeria/internal/server/thrift/ThriftDocServicePlugin.java:
```diff
@@ -495,11 +494,12 @@ public Map<String, String> loadDocStrings(Set<ServiceConfig> serviceConfigs) {
+                             .collect(toImmutableMap(Map.Entry<String, String>::getKey,
+                                                     (Map.Entry<String, String> entry) ->
+                                                             DescriptionInfo.of(entry.getValue()),
+                                                     (a, b) -> a));
```
Thanks ! I'll check it 😎

cnabro replied:
`Functions.first()` return type is `public static <T, U> BiFunction<T, U, T>`, 
But toImmutableMap only supports `public static <T> BinaryOperator<T>`.

So It can't be used and It should be changed to be used.
```
// AS IS
public static <T, U> BiFunction<T, U, T> first() {
        return (first, second) -> first;
    }

// TO BE
public static <T> BinaryOperator<T> first2() {
    return (first, second) -> first;
}
```

ikhoon replied:
I don't know that it uses `BinaryOperator`. Please ignore the comment.

cnabro commented on core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedDocServicePlugin.java:
```diff
@@ -145,7 +147,7 @@ public ServiceSpecification generateSpecification(Set<ServiceConfig> serviceConf
     }
 
-    private static void addServiceDescription(Map<Class<?>, String> serviceDescription,
+    private static void addServiceDescription(Map<Class<?>, DescriptionInfo> serviceDescription,
```
OK. I'll wait for it 👍

ikhoon commented:
Thanks a lot for your hard work! ❤️
